### ISSUE TITLE
derivative performance

### ISSUE THREAD

**Issue by firepick1** [josdejong/mathjs#811]
I've been using mathjs with great success in my Javascript neural network framework, Kinann:
https://github.com/firepick/kinann
Mathjs provides the key tool I need to perform gradient descent, namely, expression parsing.
The recent introduction of derivative has been critical--thank you so much for this feature!

I am using Kinann to write calibration software for robots such as the rotary (or revolute)  three-arm delta. The kinematic equations are a bit large. For example, here is the equation for the x world coordinate  given three angles x0,x1,x2:
https://gist.github.com/firepick1/53c2f9a698a063a191e364d745880f87
This is just one of the equations I'm working with. They are all that unwieldy.

With equations of this size, performance is always a concern. Indeed, I've created a separate Optimizer class in Kinann for factoring out common subexpressions that can be independently memoized to speed up evaluation. Each subexpression is evaluated once:
https://github.com/firepick/kinann/blob/master/src/Optimizer.js
Optimized expressions evaluate much quicker than eval() of unoptimized expressions.

Although I've solved the eval() problem, I have just run into a worse problem with derivative(). The derivative of these complex expressions does not compute after minutes of waiting. Uh. Oh.

I am currently investigating the implementation of a derivative() method for optimized expressions. This would compute and memoize the subexpression derivatives. It looks promising, but these are early days.

I have taken variious shortcuts in Optimizer implementation that would be perhaps inappropriate for mathjs. In particular, I define "subexpression" as "anything in parentheses".  This is my poor-mans' substitute for a robust expression Node optimizer. 

The larger issue is that mathjs expressions are so useful that we need to work on optimizations that can benefit everybody, not just me.

**Comment by josdejong** [comment]
Thanks for the detailed explanation of your use case. If you or anybody else has concrete ideas on how to improve performance - all ideas are welcome. Maybe we should define a simple benchmark to see what effect various ideas have on the performance, and to see whether there is a specific bottleneck which needs to be addressed.

**Comment by firepick1** [comment]
My current benchmark is the one that won't terminate on mathjs: https://gist.github.com/firepick1/53c2f9a698a063a191e364d745880f87
I have a Pixel2 Chromebook and left it running for minutes. Since I need a derivative computable in seconds, this may prove useful in gauging progress. It's a rather brutal benchmark, but very real. 

Regarding an approach for performance, I think that it is useful to decompose an expression into subexpressions. Essentially, we would have a "system of related expressions". For example, consider:
(x+1)*sin(x+1)
This expression can be broken down into:
f0 = x + 1
f1 = sin(f0)
f2 = f0 * f1

Evaluating f0,f1,f2 in turn gives us the eval() value of the expression. We can also take the derivative of each of the equations and compute the derivative at any x:

f0_dx = 1
f1_dx = cos(f0) * f0_dx
f2_dx = f0 * f1_dx + f1 * f0_dx

I'm slowly making progress and hope to report back soon on how this approach works or fails.

If this approach does work, then we would need to discuss how to tackle mathjs itself. My Optimizer class takes a lot of shortcuts and liberties that should not be allowed in mathjs. For example, Optimizer ignores imaginary numbers, and it automatically generates identifiers that might clash with user variables, etc. Also, it just seems horrible to reimplement derivative() again in Optimizer as I'm doing. That's definitely a red flag. So I consider Optimizer as a proof of concept, not the true solution we need.

**Comment by josdejong** [comment]
Thanks for the update! The benchmark would be a bit more practical if it runs for a few seconds rather then forever ;).

**Comment by firepick1** [comment]
Yes. It is quite the enfant terrible.
I also realized that you can't use the gist directly since it is multiline. I didn't realize that mathjs created multi-line expressions, so I had to do this to tackle the evil benchmark:

gist = gist.replace(/\n/g," ")

Interestingly, I am able to take the derivative of the gist in about 7ms (!) with the Optimizer class.
This is promising, but I still need to do a lot of double checking and testing. :)

**Comment by firepick1** [comment]
Jos, I have rewritten Optimizer.js as Equations.js, which runs even faster.
https://github.com/firepick/kinann/blob/dev/src/Equations.js
Equations computes the derivative of the gist in well under 20ms, which is fast enough for my neural net.

When you have a moment (well maybe a bit longer), take a look at the unit tests in the bottom half of the file to see how Equations is used. Conceptually, Equations tracks a group of named expressions (i.e., equations) as well as all their sub-expressions. Computing the derivative is fast because the derivative of each subexpression is only calculated once. Indeed, the derivatives are ALSO added to the Equations object.

Here's how you define equations:
  var eq = new Equations();
  eq.set("y", "m*x+b");
  eq.set("z", "sin(m*x+b")

Since Equations keeps track of all the equations, it may automatically return simplified expressions:

  eq.get("z") // returns  sin(y)  instead of sin(m*x+b)

In terms of integration with mathjs and resolving this thread's issue, we should probably discuss the Examples.nodeDerivative() function first.  In particular, I hope that mathjs might provide an equivalent function with appropriate callbacks. As you can see, I was forced to reimplement a substantial subset of mathjs.derivative() and I'd rather share a merged solution. 

Thanks,
Karl

**Comment by josdejong** [comment]
That sounds great. So if I understand it correctly your solution is much faster because it smartly memorizes derivatives so it only has to do this once.

I will look into your solution in more depth soon, It's not exactly clear to me yet how generic your solution is and what it would involve to integrate it in math.js itself.

**Comment by firepick1** [comment]
Equations'js is fundamentally a forest of named mathjs expression trees. Expression tree forests do address the derivative() performance issue, but they can do a lot more. Equations.js now has compile(), which returns a Javascript function that evaluates the forest for a given scope. With this new function, the performance benefits of an expression tree forest extend to eval() as well, which is significantly faster. Indeed, one might be tempted to implement...a spreadsheet using an expression tree forest. :)

var eq = new Equations();
eq.set("a1", 1);
eq.set("a2", 3);
eq.set("b1", "a1^2)
eq.set("b2", "a2^2")
var feval = eq.compile();
should.deepEqual(feval({}, {
  a1:1, b1:1,    // spreadsheet row 1
  a1:3, b2:9     // spreadsheet row 2
});

https://github.com/firepick/kinann/blob/dev/src/Equations.js

**Comment by josdejong** [comment]
From your code I see that your `simplify` is a more simple and straight replacement of inputs, not via a comparison against a set of rules like the `simplify` of mathjs does. I suppose that makes the version of mathjs much slower. But isn't your version more limited in the cases that math.js handles? It looks like it has much less rules built in.

I'm not sure what you mean with an "expression tree forest" and I didn't immediately spot how/why your version of derivative is so much faster (besides that it looks simpler and more direct which is always good :).

**Comment by firepick1** [comment]
Thanks for reading the code and noticing fastSimplify. The fastSimplify() method is much faster than simplify but only applies identity simplification and ellminates 0 and 1 where possible. It doesn't do as nice a job as mathjs simplify, which applies many more rules. But fastSimplify() does reduce computational burden, so it has benefit as an implementation for performance. The equations I am working with are nasty and I never really want to look at them--I just want them to evaluate quickly. If simplify() had an option to do fastSimplify(), I could switch back to using simplify.  If you'd like, I can fork mathjs and send a pull request for that. 

Although fastSimplify() does improve performance, it's not the primary factor. The key to performance is that we split up a large mathjs parse tree into named subtrees. We also cache (or memoize) these subtrees for future reference. By caching the subexpressions, we only have to take the derivative of a repeated subexpression once. For example, in the gist, you probably noticed a lot of (f - e) expressions that are now only processed once.  Equations uses a digestNode() function to break down a large parse tree into smaller subtrees.

Consider the equation "y = (1-x^2) * x^2". Notice that it has a repeated x^2, We digest the expression as a collection (or forest) of named subtrees:
_0 = x^2
_1 = 1-_0
_2 = _1 * _0
y = _2

After digesting the original equation for "y", we now have a total of four equations. The variables _0, _1 and _2 are automatically generated by Equations.  Notice that these equations are very simple and we can use them to compute our derivative, "y_dx":

_3: "2 * x ^ 1"
_0_dx: "_3"
_1_dx: "- _0_dx"
_2_dx: "_1 * 0_dx + _0 * 1_dx"
_y_dx: _2_dx

Equations keeps a two-way lookup map from names to subexpressions and from subexpressions to symbols. This means that any parse tree that has "x ^ 2" can replace that subtree with "_0".  It also means that the more equations we add to Equations, the faster it performs, especially if there are common subexpressions such as "x ^ 2".

For small equations, this approach is overkill. But for large expressions and multiple expressions, it is quite fast and efficient.

**Comment by firepick1** [comment]
I ran a benchmark that can help explain how Equations handles the gist quickly. First let's parse the gist using mathjs to give us a timing baseline:
mathjs.parse: 16ms

Once we have the parsed node tree, we see that fastSimplify runs quickly on the parsed tree, which proves that the mathjs expression tree is very efficient for expression manipulation:
Equations.fastSimplify: 5ms

The set() method is used to digest a new equation. It is a superset of mathjs.parse and also memoizes the subtrees. It therefore is the workhorse of Equations. Notice that memoization takes longer than parsing:
Equations.set: 35ms

Interestingly, memoization introduces another consideration in that we incur a burden on reconstructing digested equations. Equations does not store the original equations--it just stores digested equations. 
Equations.get: 32ms

The surprise comes when we time the derivative itself. Notice that derivative() is faster than parse() or set(). This highlights the benefits of digesting the parse tree into memoized subexpressions
Equations.derivative: 10ms

For details, see line 153 of https://github.com/firepick/kinann/blob/master/test/Equations.js

**Comment by josdejong** [comment]
@firepick1 sorry for my late reply, was burdened in a bit too much work. 

So if I understand it correctly, `fastSimplify` currently only handles a subset of the rules that `math.simplify` currently supports. I suppose though that it can be extended to support everything that's currently in `math.simplify`?

If I'm not mistaken the "memoization" part is the results cached in `symbolTreeMap`? As far as I can see there is currently no limitation there, so memory usage can grow infinitely. I suppose there should be some limit there (using an aging hash map or something like that)?

Your solution looks simple and promising. Would you be interested to work it out in more detail in order to replace the current implementation? Or alternatively, introduce a caching solution in the current implementation to speed it up?

**Comment by firepick1** [comment]
Jos, thank you for the opportunity to contribute to mathjs. I shall send you a pull request when I am ready. :)
It will take some time, since this is delicate surgery.

**Comment by josdejong** [comment]
Great, thanks beforehand! Please keep communicating whilst you're working on it. I expect there may arise difficulties in keeping the behavior the same as the current implementation. Just add as much unit tests as needed to ensure the desired behavior :)

**Comment by tetslee** [comment]
It is possible that `simplify` was getting in an endless loop with the gist benchmark. 

I've changed the terminate condition in #848 to avoid endless loops. It's still really slow running that benchmark but at least it eventually terminates.

**Comment by firepick1** [comment]
I've been looking at the existing simplify in mathjs and it occurs to me that the least invasive way to incorporate Equations.fastSimplify into mathjs is to simply pass it in as a rules function per the existing API. This is completely backwards compatible yet offers performance as needed. I see that simplify itself has default rules. We could add a method simplify.fastSimplify so that folks could write this:

mathjs.simplify("a+(b+0)*(c*1)", [mathjs.simplify.fastSimplify]);

Is this acceptable? If so I can proceed directly to implementation in my develop branch of my fork and submit a pull request.

**Comment by josdejong** [comment]
That would be a great start! Start from that and piece by piece replace the old functionality with the new solution. Good idea. Whilst the "migration" is in progress I would like to keep it in a separate branch (can be your current branch) though and not release it when half way the process. 

If things are not exactly compatible we can discuss the options - with simplification rules there is not always a single "best" solution, there may be trade offs worth discussing.

**Comment by firepick1** [comment]
I realized I could add fastSimplify to: 1) the beginning of simplify.rules or 2) the end of simplify.rules. These two testing strategies allowed me to directly compare fastSImplify with simplify. Along the way I found some bugs in fastSimplify (i.e. BigNumbers) and then I ran into something that requires discussion.

There is an interesting test case. "3+2/4"
Existing simplify changes this to "7/2"
fastSimplify changes this to "3.5"
Notice that for the test case: "3+1/3", both simplify and fastSImplify return "10/3"

Are you OK with 3.5 or do you require 7/2?

**Comment by josdejong** [comment]
Can you explain when fastSimplify will output a number `3.5` and when a fraction like `7/2`?

**Comment by firepick1** [comment]
The heuristic is designed to preserve accuracy while reducing fractions with denominators of 1,2,4,5 and 10.  The numbers 1 and 2 are particularly important given that they often show up as by-products of  common operations. The numbers 4,5,10 were added as a "well why not" case given that they tend to yield "simpler" results. Note that no attempt was made to cover all such denominators. Rational numbers are perfectly accurate but annoying in practice--it's easier to deal with 0.5 than 1/2 in an expression. This is where an engineer and a mathematician might disagree. Hence the discussion. I'm an engineer. Here are the options in my order of preference:

OPTION A: adopt fastSimplify heuristic
OPTION B: preserve simplify behavior
OPTION C: add a rule to specify array of "collapsing denominators"

**Comment by HarrySarson** [comment]
Is it possible that division by 2 followed by division by 3 will produce an irrational number. If fastSimplify converts `1/2` to 0.5 and then `0.5/3 = 0.16667` then accuracy is lost.

**Comment by josdejong** [comment]
@firepick1 that sounds good. I totally agree that there are heuristics at play here, your approach sounds good to me. We probably need fine-tuning in the future and that's just fine.

@HarrySarson that's a good point. So basically you would like to keep integer numbers as fractions (or any number that can be represented as fraction without losing accuracy) during the simplifying process, and only at the end decide once whether to turn the fraction into a floating point number or not.

**Comment by HarrySarson** [comment]
I guess as far as the implementation details are concerned there may be no difference between `1/2` and `0.5` but the external API should be consistent.

 - `math.divide(math.simplify('3 + 1/2'), 3)` should definitely give `math.fraction(7,6)`.
 - `math.divide(math.simplify('3.5), 3)` should definitely give `1.16667`.

I haven't looked at the simplify function implementation so this may already be the case but I feel like the above would be a good to add to the unit tests - especially if there is a danger of mixing rational and floating point numbers.

**Comment by firepick1** [comment]
We all agree on the first.

However, I would argue that the second one should return the fraction '3.5/3' simply because no precision is lost. Dividing by 3 loses precision and we can never get it back. Simplify should "do no harm".

**Comment by josdejong** [comment]
> Saying that the current behaviour is to return 7/6 in both cases.

that would be nicest indeed. 

> Simplify should "do no harm".

definitely!

**Comment by firepick1** [comment]
Jos, I created a simplify branch in my fork with the following changes:
https://github.com/firepick1/mathjs/commit/ba4f0670571ee42825a27405b68813a5829d7c28

Things to notice are:
1) simplify computation cost is proportional to the number of rules, since each rule completes an entire expression tree pass. Reducing rules speeds up simplify.
2) fastSimplify eliminates a few rules
3) I commented out rules that have no tests. :)
4) I added new tests for identity removal--these tests revealed some cases not handled by existing rules. The fastSimplify code deals with those test cases.
5) the existing test suite found bugs in fastSimplify which I have fixed. :)

Are we ready for a pull request?

**Comment by josdejong** [comment]
That's good news! So now the "hybrid" solution is up and running :)

> Are we ready for a pull request?

We can merge your work in a separate branch in the mathjs project, but I really would like to merge it into develop once the old solution is completely replaced by fastSimplify (see https://github.com/josdejong/mathjs/issues/811#issuecomment-302495073).

**Comment by firepick1** [comment]
The fastSimplify implementation does not handle the collection of like terms nor the simplification of constant expressions using fractions. Interestingly, the downstream fastDerivative code doesn't actually require that level of simplification. The reason for this is that the fastDerivative memoization computes sub-expressions once and reuses the results for duplicated subexpressions. Expending more computation to collect like terms and reduce constant expressions would actually be less beneficial for certain applications. Indeed, I think we have conflicting requirements:

1) fast simplification => the evaluation of the simplified expression is more important that the symbolic form
2) terse simplification => the symbolic form of the simplified expression is more important than its evaluation

These conflicting requirements actually represent separate use cases. Fast simplification is important for applications such as gradient descent. Terse simplification is most critical for human comprehension and communication. Both are requirements are important on their own. However, merging them may be very difficult.

Interestingly, the existing rules architecture of simplify neatly provides a framework rich enough to handle both use cases--but with different sets of rules:
1) simplify(expr, [fastSimplify])  => fast simplification for computation
2) simplify(expr, [fastSImplify, existing_rules) => terse simplification as per existing production mathjs
3) simplify(expr, [fastSimplify, super_duper_new_rules]) => fantastically beautiful terse simplification

In other words, I really like the mathjs approach of simplification rules. The rules provide a very precise way of supporting multiple simplification paradigms. I actually would not add more code to fastSimplify because it would make a simple function very complicated. But I do think we should add more functions such as "combineLikeTerms" and "simplifyConstantV2" to perform powerful simplifications in a single pass. Doing so would improve overall performance by reducing the number of tree passes and would also provide opportunity for future improvements.

**Comment by josdejong** [comment]
Thanks, that makes sense to recognize these two different use cases (fast computation vs. best simplified expression).

I'm not sure whether these two conflicting requirements require two different implementations. Would it be possible to have a single implementation with an option for the "depth" to which it simplifies the input?

From a maintenance point of _view_ I'm very careful having to maintain two similar solutions, that will be really a pain. I strongly prefer having a single implementation. 

My feeling is that the current (slow) implementation has a lot of room for performance improvements, by smartly applying some of the techniques you use in `fastSimplify`. And it can probably be extended with an option to do a fast, rough simplification or the best possible. Or the other way around - extend `fastSimplify` to meet the functionality the current `simplify` offers (with an option to go for fast or thorough).

You're more familiar with the code than me, do you think there are options there?

**Comment by firepick1** [comment]
Ahhhh. Ok. I still don't fully understand the existing simplify code myself, but what I see suggests that the reason it is slow is that it runs through 34 separate default simplification rules. For example simplifyConstant is just one rule. There are 33 others. By introducing fastSimplify, I have reduced the list of rules to 23. So it's not really a difference in algorithms. fastSimplify essentially combines about 6 rules. Then there were rules that weren't tested, so i commented them out. I am actually thinking of renaming fastSImplify to "simplifyIdentities", which reduces a lot of the junk created by derivatives. This will allow us to make incremental progress in improving simplify. Each simplify module can then become its own thing and folks can combine them as they wish in different ways. For example, somebody can code up a "combine like terms" module for simplify. Doing that would probably double the speed of simplify because most of the rules now deal with combining like terms. If we combined like terms in a single pass, then it would be faster. Some of us would need such a feature, and others like me don't need that default feature. Another module that would need to be recoded is simplifyConstant. Right now it is kinda complicated and probably should make better use of Rational as you have suggested elsewhere.

I will go ahead and merge my simplify changes into the develop branch and submit a pull request. I think my changes will be compatible with the original simplify architecture since all I'm doing is collapsing multiple rules into a single one-pass rule. We have more work ahead of us but this is a baby step.

**Comment by josdejong** [comment]
I think `fastSimplify` is faster due to other reasons than having less rules: it has a very simple and cheap loop which checks what the type of node is now and what to do. On the other hand, `simplify` interprets rules, loops over rules, matches them, creates "rule match results", interprets that again, etc. There is simply a lot more work done each loop, and I think there are ways to improve the existing implementation drastically.

Before we can say something meaningful about that however, we should have a useful benchmark (something that touches most functionality of `simplify` and which runs for say max 1 minute), then we can do some profiling to see where the current bottleneck is and see how to improve it.

So as for concrete action points:
- Create a benchmark for this in the `develop` branch of mathjs, so we have something concrete to compare against.
- Create a separate `simplify_v2` branch to work on merging the `fastSimplify` and `simplify` implementations further until it's a coherent whole instead of two different implementations. @firepick1 
 would you like to become a collaborator on `mathjs` so you can easily work on this separate branch? Or do you prefer working in your own fork of mathjs?

**Comment by josdejong** [comment]
@firepick1 I've created two simple benchmarks in the `develop` branch to see how good both `simplify` and `derivative` do. We can use this to see how good any improvement does.

See: https://github.com/josdejong/mathjs/blob/develop/benchmark/algebra.js

On my computer it outputs:
```
simplify 2 * 1 * x ^ (2 - 1)
    2 * x
derivative 2x^2 + log(3x) + 2x + 3
    4 * x + 3 / (3 * x) + 2
algebra simplify                         x 826 ops/sec ±0.87% (89 runs sampled)
algebra derivative                       x 228 ops/sec ±0.27% (87 runs sampled)
```

**Comment by firepick1** [comment]
Jos, I think it would be more efficient to work as a collaborator in a separate branch as you suggest. This will save me the hassle of upstream pulls and pull requests. And I like having a simplify2 branch to coordinate that long-lived set of feature improvements for simplify--there's a lot of work that can be done in simplify by many folks. I will confine my changes to simplify2 and consult with you on proposed merges into develop via this thread so that the community can weigh in. Thank you for extending the collaboration proposal--it is quite a challenge managing multiple contributors and I appreciate and respect the trust implied.  Also many thanks for the benchmarks. I may add more as we proceed.

**Comment by josdejong** [comment]
You're welcome! I've added you as a collaborator. I've seen you can build solid and well tested code and you're open in communication. I trust you will be careful with being a collaborator. Can you please email me your email address in case that's needed some day? And if you have any questions about how stuff works just ask me.

You should be able to create a `simplify2` branch now and get going :)

**Comment by firepick1** [comment]
UPDATE: I've created a simplify2 branch and added a new function simplify.simplifyCore(), which consolidates many existing simplifications into a single function. This one function yields about a 75% performance improvement in ops/sec (@Jos thanks for the benchmark!). All tests pass and I added a few more, so I'd say this commit is releasable at your discretion (but you might want to wait till I finish up the variable resolve feature request).
[benchmark.txt](https://github.com/josdejong/mathjs/files/1191586/benchmark.txt)

Interestingly, this change ALSO speeds up derivative(), but not enough to close this feature request. We still need to figure out the best way to integrate the equation memoizer into mathjs and I currently have no proposal to put forward just yet.

**Comment by josdejong** [comment]
Thanks for the update, sounds good! I will check out your work this weekend to see where we are now.

**Comment by josdejong** [comment]
I've checked out the `simplify2` branch. It's really neatly structured with `simplifyCore` being an (internal) part of `simplify`, thanks for all your effort @firepick1.

Whilst I'm still not happy having to support two different implementations of `simplify` at the same time, I think it's worth releasing it now since it offers such a huge performance improvement (which is what counts for the user), and I've just merged `simplify2` into `develop`. I hope in the end we can move all built-in rules from `simplify` to `simplifyCore` though!

I've replaced some ES6 usage (const and arrow functions, backticks) with old fashioned ES5. We don't yet transpile the source code, so this fails on old versions of node.js and old browsers. It's time though to set up build scripts such that we _can_ use ES6...

**Comment by firepick1** [comment]
Oops. Sorry about the ES6 infection. I've been very spoiled. I tell all my customers to require Chrome and not bother with Internet Explorer--Microsoft will catch up eventually.

And I'll continue thinking about ways to incorporate more into simplifyCore--it's tough because they serve different purposes. The legacy simplify() is exhaustive in its multi-pass search for canonical simplification. In contrast, simplifyCore() executes in a single pass for performance. Brutal expedience vs. elegant refinement.

**Comment by josdejong** [comment]
Thanks again Karl! Maybe you're right and we can't unify this. Let's just keep it in the back of our minds.

**Comment by mollthecoder** [comment]
It's been 3 years and this still isn't closed. Might you want to close it?

**Comment by firepick1** [comment]
Ah. Sorry. I forgot I created the issue and didn't close it after I fixed it.

**Comment by firepick1** [comment]
NOTE: I've been using tensorflowjs lately and their solution is quite extensive. Tensorflow relies on sophisticated memoization as well as GPU acceleration if available. Tensorflow does present a steep learning curve, however. For example, data is given asynchronously with the data() method of a tensor whereas in Mathjs, we can use arrays directly.
